### ISSUE TITLE
Memory increasing after event-handlers merge

### ISSUE THREAD

**Issue by cacheworks** [meetecho/janus-gateway#727]
Hello

Hope you all had good holidays 👍 

After running Janus for a while I decided to update to the refcount branch to see if that would get rid of the occasional race condition.

Below examples are with events enabled and only the videoroom plugin being used 

After switching to the refcount branch (commit fac6fad0105fcb2ff430beb86f57ae37f1ff1cf2 ) memory kept increasing (without AddressSanitizer this time)
![image](https://cloud.githubusercontent.com/assets/13921000/21590281/b97c1c54-d0fa-11e6-9bd3-bf736dbb9dab.png)

So I switched to the master branch to see if the same happens there and it does (commit e4d01db53618102ef9ca7ec63d35a2c8ade1bb3b ) (I killed the process so the swap didn't grow as large this time)

![image](https://cloud.githubusercontent.com/assets/13921000/21590381/a74db3fc-d0fb-11e6-8a71-23a2b2912c65.png)

I then reverted back to commit e2a8e496ffc8ceb3057b49c93e037e85bcff8d54 to see if the issue existed before the event-handler branch got merged and it appears the issue doesn't appear this time

![image](https://cloud.githubusercontent.com/assets/13921000/21590412/dbb02fd0-d0fb-11e6-82b7-eb48975e78da.png)

It seems to me that the issue is caused by 6a998158ab0137ef2faa5bb377902bedf0f7e089 and was not (fully) solved in eba63a1926ea9f0812123cc110c13a62547a9147

If I can help you spot the problem by trying out other commits let me know and I'll get those deployed

Thanks for your support, it's much appreciated

**Comment by lminiero** [comment]
Not fully back from vacation, actually, that will only happen in a few days.

Anyway, my guess is that there are some events we're not freeing: some I fixed in the commit you mentioned, but there may be more. One thing you can check is whether that happens only when event handlers are enabled or not: if you're using the VideoRoom, you can also try to only disable VideoRoom events (from its configuration file) to see if the leaks happen there or not.

**Comment by lminiero** [comment]
As a side note, what are you using for monitoring and tracing those graphs? Just to see if we can replicate this in our labs somehow as well.

**Comment by cacheworks** [comment]
I'm using [Munin](http://munin-monitoring.org/) to plot these graphs.

I had a chance to restart the gateway to try disabling the events like you suggested.

I used commit eba63a1926ea9f0812123cc110c13a62547a9147 and disabled the event broadcasting in janus.cfg, memory did not increase that day
![image](https://cloud.githubusercontent.com/assets/13921000/21639831/8557a8ac-d272-11e6-8ca3-bca01c0e821e.png)

I then enabled the event broadcasting in ```janus.cfg``` but disabled the videoroom from sending events in ```janus.plugin.videoroom.cfg```, which caused it to increase memory again; (perhaps a bit slower though?)
![image](https://cloud.githubusercontent.com/assets/13921000/21639979/4a4898ba-d273-11e6-932c-146011716de1.png)

It seems that it started to increase faster as more publishers and subscribers joined

**Comment by lminiero** [comment]
Ok, if disabling events doesn't result in memory increasing, then it means events are definitely to blame. The fact that memory only increased slightly when events were enabled but those from the VideoRoom were not, can be blamed on the fact that some events (in particular, all those related to the life cycle of a PeerConnection) are still originated on a regular basis by the core anyway. It might be harder to spot exactly which ones are not freed correctly (probably a matter of how we're constructing them, though), while it should be easier for the VideoRoom as it's an isolated component. I'll have a look and come back to you when I have some news.

**Comment by lminiero** [comment]
Did you not compile the sample event handler, or is it disabled via configuration file? I just fixed a leak that may be caused by events enabled but missing handlers: basically since the mechanism is enabled, the event gets to the `janus_events_notify_handlers` method, but in its current stage if no event handler is available (for one reason or another) the method immediately returns. This can cause a leak for some events, especially those that include an allocated `json_t` object as part of the info they provide (WebRTC, media, core, plugins). I've modified the code to find and free those in case there's no handler to notify.

If you can confirm that's your scenario, I'll push the fix so that you can test it.

**Comment by lminiero** [comment]
Can you test this and let me know if that helps?

**Comment by cacheworks** [comment]
I compiled  with the sample event handler enabled and setting ```broadcast=no```` in ```janus.cfg``` seems enough to stop it from increasing memory.

When I enable the events to be broadcasted I use the sample plugin without any changes and have a backend receiving them and I don't see any time outs occuring

Do I understand correctly that your commit should fix a memory leak when ```broadcast = yes``` and ``` disable = libjanus_sampleevh.so```, that is not a scenario I've tried yet. I'll happily try it out though. which configuration would you suggest to try?

Thank you for taking a look 👍

**Comment by lminiero** [comment]
It should fix it for when you have `broadcast=yes` in `janus.cfg`, but `enable=no` in `janus.eventhandler.sampleevh.cfg`, which should be pretty much the same as `disable=libjanus_sampleevh.so` as the plugin disables itself. The `enable=no` in `janus.eventhandler.sampleevh.cfg` is the default when you install Janus, so if you just enabled it in `janus.cfg` you may have been getting that leak.

**Comment by lminiero** [comment]
If instead the module was loaded and enabled for you, the leak is likely somewhere else.

**Comment by cacheworks** [comment]
Ok thanks, to confirm I do have ```enabled = yes``` set in ```janus.eventhandler.sampleevh.cfg``` and am receiving events so it seems like it wouldn't enter that if statement when the leak shows itself.

Could cd96a5dd0d34414a34ec3746b5ffe272d449c1e7 perhaps be related? with the amount of selected-pair events I could see curl events adding up quickly so that seems worth a try 👍 

If not, would it help to try all the different values for ```events``` in ```janus.eventhandler.sampleevh.cfg```
It might take a while to try them all but it could narrow down which (or all) events cause this?

**Comment by lminiero** [comment]
That may be it: I noticed those libcurl leaks in my AddressSanitizer log and so fixed them as well. No need to narrow the events down, don't worry: if possible, just check if the two leaks I fixed do help your memory issue, and in the meanwhile I'll keep looking for places where we don't decref in case it doesn't.

**Comment by cacheworks** [comment]
I've been running cd96a5dd0d34414a34ec3746b5ffe272d449c1e7 and it has definitely fixed the issue that caused the memory to increase initially

![image](https://cloud.githubusercontent.com/assets/13921000/21767726/f8a86a06-d675-11e6-81b1-893fd246d0ce.png)

The first increase in this graph shows the master branch before cd96a5dd0d34414a34ec3746b5ffe272d449c1e7 got merged.

After that it ran for a few days without linear increase but there were a few bumps, which I was confused by until it suddenly increased a lot again and crashed (twice)
After looking into this I I noticed the event handler nodejs process was down, starting that stopped the memory from increasing again.

So to sum up cd96a5dd0d34414a34ec3746b5ffe272d449c1e7 fixed the issue I was initally having but another memory leak seems to appear when the backend for the sample event handler is down (I'll see if I can get any more useful info here and submit it as separate issue)

Thanks a lot for solving this issue 👍

**Comment by lminiero** [comment]
Thanks for the detailed report as usual!

As to the nodejs service being down, in that case memory should indeed increase as we keep all the events in queue: in fact, a failure to transmit will result in retransmissions following an exponential backoff mechanism, and a retransmit of the same event/group of events is attempted again and again keeping the other packets in queue. If all retransmissions fail (5 times by default), then we give up and read the next events in queue. So in your case it makes sense that memory was growing, as the queue was being filled with events that were not being sent.

That said, those events should have been then freed after the last retransmission failures, and at the same time restoring the backend should have resulted in the memory going back to normal again as the pending messages in the large queue should have all been sent normally. If this was not happening (that is, restoring the backend prevented the memory from growing further, but the memory that was added in the meantime was never freed) then it's indeed a leak I'll have to look into. Your detailed analysis definitely allows for a drill-down investigation, so if there is something to fix it should be easy to spot.

**Comment by lminiero** [comment]
I can't find any leaks there. My guess is that in your case the memory increase never re-entered due to the crash that followed: had the web backend been restarted sooner, my guess is that all the memory would have been freed again, as from what I can see in the code we're doing the right thing.

Why it crashed is something else to look into (maybe too many packets being enqueued and never served?). Is there any debugging data you can share on that? Should that happen again, feel free to open a new issue just for that.
